ambry-replication/src/main/java/com/github/ambry/replication/PartitionLeaderInfo.java

```diff
@@ -0,0 +1,155 @@
+
+import com.github.ambry.clustermap.PartitionId;
+import com.github.ambry.clustermap.ReplicaId;
+import com.github.ambry.clustermap.ReplicaState;
+import com.github.ambry.server.StoreManager;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.locks.ReadWriteLock;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import static com.github.ambry.clustermap.ClusterMapUtils.*;
+
+
+/**
+ * Maintains the list of leader partitions on local node and their corresponding peer leaders in remote data centers
+ */
+public class PartitionLeaderInfo {
+
+  private final Map<String, Set<ReplicaId>> peerLeaderReplicasByPartition = new ConcurrentHashMap<>();
+  private final StoreManager storeManager;
+  private final ReadWriteLock rwLock = new ReentrantReadWriteLock();
+  private static final Logger logger = LoggerFactory.getLogger(PartitionLeaderInfo.class);
+
+  public PartitionLeaderInfo(StoreManager storeManager) {
+    this.storeManager = storeManager;
+
+    // We can't initialize the peerLeaderReplicasByPartition here because we don't know the leader partitions on local node (server) until it has finished participating with Helix.
+    // peerLeaderReplicasByPartition map will be updated after server participates with Helix and state of replicas transition to LEADER (via onPartitionBecomeLeaderFromStandby())
+  }
+
+  /**
+   * Get a map of partitions to their sets of peer leader replicas (this method is only by ReplicationTest for now)
+   * @return an unmodifiable map of peer leader replicas stored by partition {@link PartitionLeaderInfo#peerLeaderReplicasByPartition}
+   */
+  public Map<String, Set<ReplicaId>> getPeerLeaderReplicasByPartition() {
+    return Collections.unmodifiableMap(peerLeaderReplicasByPartition);
+  }
+
+  /**
+   * Add a leader partition and its set of peer leader replicas. This method is thread safe.
+   * @param partitionName name of the partition to be added
+   */
+  public void addPartition(String partitionName) {
+
+    // 1. get local replica from store manager
+    ReplicaId localReplica = storeManager.getReplica(partitionName);
+
+    // Read-write lock avoids contention from threads removing old leader partitions (removePartition()) and threads updating existing leader partitions (refreshPeerLeadersForAllPartitions())
+    rwLock.writeLock().lock();
+    try {
+      // 2. Get the peer leader replicas from all data centers for this partition
+      List<? extends ReplicaId> leaderReplicas =
+          localReplica.getPartitionId().getReplicaIdsByState(ReplicaState.LEADER, null);
+
+      // 3. Log the list of leader replicas associated with this partition (will be used later for leadership based replication)
+      List<ReplicaId> peerLeaderReplicas = new ArrayList<>();
```

Comment:
How about making `peerLeaderReplicas` into a HashSet so that it doesn't have to be copied into a set on line 88?